cmd/whclient/main.go

```diff
@@ -0,0 +1,204 @@
+package main
+
+import (
+	"io"
+	"net"
+	"os"
+	"os/signal"
+	"strconv"
+	"sync"
+	"syscall"
+	"time"
+
+	set "github.com/deckarep/golang-set"
+	"github.com/docopt/docopt-go"
+	log "github.com/sirupsen/logrus"
+	"github.com/taskcluster/taskcluster-client-go"
+	"github.com/taskcluster/taskcluster-client-go/auth"
+	"github.com/taskcluster/webhooktunnel/whclient"
+)
+
+const usage = `Webhooktunnel Client
+Webhooktunnel Client is a command line utility which establishes a connection
+to the webhooktunnel proxy and allows serving http without exposing ports to 
+the internet.
+
+[Firewall/NAT [User] <--]---> [Proxy] <--- [Web]
+
+Usage: whclient <clientID> <accessToken> <targetPort>
+whclient -h | --help
+
+Options:
+-h --help Show help`
+
+const closeWait = 2 * time.Second
+
+func main() {
+	log.SetFormatter(&log.JSONFormatter{})
+
+	arguments, _ := docopt.Parse(usage, nil, true, "Webhook Client 0.1", false)
+
+	clientID, accessToken := arguments["<clientID>"].(string), arguments["<accessToken>"].(string)
+	targetPort, err := strconv.Atoi(arguments["<targetPort>"].(string))
+	if err != nil || targetPort < 0 || targetPort > 65535 {
+		log.Fatal(usage)
+	}
+	// Configure signals for graceful handling
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+
+	// hold count of running connections
+	var running sync.WaitGroup
+	// used for set operations
+	count := uint64(0)
+	runSet := set.NewSet()
+	run := true
+	// accept new streams from this channel
+	strChan := make(chan net.Conn, 1)
+
+	client, err := whclient.New(makeConfigurer(clientID, accessToken))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	go func() {
+		for {
+			// This simulates http.Server's Serve method
+			waitTime := 5 * time.Millisecond // default wait time when connection fails
+			stream, err := client.Accept()
+			if err != nil {
+				ne, ok := err.(net.Error)
+				if !ok || !ne.Temporary() {
+					log.Fatal(err)
+				}
+				// wait if temporary
+				time.Sleep(waitTime)
+				waitTime = 2 * waitTime
+				continue
+			}
+			// reset wait time
+			log.Printf("new connection")
+			waitTime = 5 * time.Millisecond
+
+			select {
+			case strChan <- stream:
+			default:
+			}
+		}
+	}()
+
+	defer func() {
+		_ = client.Close()
+		close(strChan)
+		close(sigChan)
+	}()
+
+	log.WithFields(log.Fields{"url": client.URL()}).Info("connected to proxy")
+
+	for run {
+		select {
+		case <-sigChan:
+			run = false
+			done := make(chan struct{})
+			go func() {
+				running.Wait()
+				close(done)
+			}()
+			select {
+			case <-time.After(closeWait):
+				// kill everything
+				it := runSet.Iterator()
+				for elem := range it.C {
```

Comment:
Instead of doing this... I suggest that we sleep until all connections are closed or if more than 5s pass we just hard exit... Calling close on everything probably doesn't do much, since exiting hard has the same effect (i Hope)..

Note: this looks good... I'm thinking we can do simpler...